### ISSUE TITLE
Btrfs-check does not check last sector accessible

### ISSUE THREAD

**Issue by socketpair** [kdave/btrfs-progs#504]
After some experiments, I accidentally made block device slightly smaller than it was when BTRFS was formatted.

I tried to mount, failed:

```
device total_bytes should be at most XXXXXX but found YYYYY
failed to read chunk tree -22
open_ctree failed
```

`btrfs check` found nothing.
`btrfs rescue fix-device-size` found nothing
`btrfs filesystem resize` requires mounted FS, so there is no easy way to fix the problem.

The device was 16M bytes less than should be. Actually, no useful data was in this tail (I mean BTRFS did not use them, so resize (shrink FS) would finish immediately).

I can reproduce the bug easily, if required.

**Comment by socketpair** [comment]
https://bugzilla.kernel.org/show_bug.cgi?id=216409

**Comment by adam900710** [comment]
It's indeed a bug in btrfs-check and fix-dev-size.

I'll enhance them pretty soon.

**Comment by socketpair** [comment]
1. Will it correct something in order to mount?
2. what if block device was truncated in a way it loses some important data?
3. What if some pointers to data become outside of the block device ?

**Comment by adam900710** [comment]
For 1, yes as long as there is no data in the truncated range.

For 2. then no way to repair without losing data, and that enhanced "btrfs rescue fix-device-size" will only error out saying the same thing.

For 3, "btrfs rescue fix-device-size" will detect and refuse to do anything.

For that case in 3), there is really not much we can do.

**Comment by socketpair** [comment]
Well, Is there any option to fix FS in case some part is not accessible anymore with some data lose ? I think solution should be the same as resolving bad sector issue.

**Comment by adam900710** [comment]
If some dev extents (can be some critical metadata) are inside the shrunk range, it's really hard to say, mostly depending on what's lost in that range.

If critical metadata, then it may fail to mount anyway.

If just some data, then it may just be some missing data.

The "btrfs rescue" method I took is really the safest method. I can also enhance the kernel, to utilize some "recue=" mount option, so that kernel will just try its best to read whatever it can get, and hope for the best luck.

**Comment by socketpair** [comment]
I think, ability to mount FS read-only and copy-out files that still readable - is a good idea.

**Comment by socketpair** [comment]
Is issue closed now ?

**Comment by kdave** [comment]
Yes, the fix has been released in 5.19.1.
